fix(cloudflare): infinite loop with more than 50 custom hostnames

mrozentsvayg opened the pull request:
…number of custom hostnames more than 50

<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

<!-- Please provide a summary of the change here. -->

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #ISSUE

**Checklist**

- [x] Unit tests updated
- [x] End user documentation updated

mrozentsvayg commented:
This one is important to include with 0.16.1 if possible, I believe.

Not currently implemented paging support for `CustomHostnames()` tests implementation contributed into not detecting this bug timely.
I'm going to add it, but we might want to merge this PR sooner than that.

mloiseleur commented:
@mrozentsvayg Do you think you can add a test on this, so we can avoid to face it again in the future ?

mloiseleur commented:
/retitle fix(cloudflare): infinite loop with more than 50 custom hostnames

mloiseleur commented:
@mrozentsvayg How about adding this known limit in the documentation ?

ivankatliarchuk commented:
Tests are missing. I have a similar use case, where need to test for 100+ zones. Here is how I did it, worth to duplicate for cloudflare? https://github.com/kubernetes-sigs/external-dns/blob/master/provider/aws/aws_fixtures_test.go

mrozentsvayg commented:
> @mrozentsvayg How about adding this known limit in the documentation ?

Do you mean hardcoded 50 results per-page in the cloudflare-go v0 api that we are currently using?
I can add it if you think it would be useful.

By the way, are there any known blockers preventing migration to a newer api ?

mrozentsvayg commented:
> Tests are missing. I have a similar use case, where need to test for 100+ zones. Here is how I did it, worth to duplicate for cloudflare? https://github.com/kubernetes-sigs/external-dns/blob/master/provider/aws/aws_fixtures_test.go

Test is easy, I drafted it yesterday. Implementing the paging specifically for the test (https://github.com/kubernetes-sigs/external-dns/blob/master/provider/cloudflare/cloudflare_test.go#L262-L282) is what I didn't yesterday have time for.
I'll do it today, hopefully there's still time to push it before the next release.

ivankatliarchuk commented:
> By the way, are there any known blockers preventing migration to a newer api ?

I think we should have at least a single stable cloudflare release first. And resolve this 2 issues
- https://github.com/kubernetes-sigs/external-dns/issues/5118
- https://github.com/kubernetes-sigs/external-dns/issues/5035

ivankatliarchuk commented:
> Do you mean hardcoded 50 results per-page in the cloudflare-go v0 api that we are currently using? I can add it if you think it would be useful.

It sounds useful. Basically we do not want to test it manually every time, and having a unit tests for paging shoudl help at least a bit. Plus as you mention, if planning to bump an SDK version,  we should have a coverage for such things

mrozentsvayg commented:
- added pagination support for custom hostnames tests
- added test for paging, it would _correctly_ hang with before-the-fix code 
- added information about custom hostnames page size to the doc

ivankatliarchuk commented:
/label tide/merge-method-squash

ivankatliarchuk commented:
Have you had a chance to test changes against a real cloudflare account?

ivankatliarchuk commented:
/ok-to-test

mrozentsvayg commented:
> Have you had a chance to test changes against a real cloudflare account?

I am running this exact build in production with hundreds of Cloudflare records and custom hostnames.

ivankatliarchuk reviewed (approved):
/lgtm

mloiseleur reviewed (approved):
342 custom hostname seems a reasonable amount to test: )
